wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+		 *
+		 * This means that "\%" behaves like "\\%", and "\_" behaves like "\\_".
+		 * To preserve this behavior, we need to add a second backslash in cases
+		 * where only one is used. To do so correctly, we need to:
+		 *
+		 *  1. Skip all double backslash patterns (as "\\" resolves to "\").
+		 *  2. Add an extra backslash when "\%" or "\_" follows right after.
+		 *
+		 * This may be related to: https://bugs.mysql.com/bug.php?id=84118
+		 */
+		$value = preg_replace( '/(^|[^\\\\](?:\\\\{2}))*(\\\\[%_])/', '$1\\\\$2', $value );
+
+		/*
+		 * 3. Unescape quotes within the string.
+		 */
+		$value = str_replace( $quote . $quote, $quote, $value );
+
+		/*
+		 * 4. Unescape C-style escape sequences.
+		 *
+		 * MySQL string literals are represented using C-style encoded strings,
+		 * but SQLite doesn't support such escaping.
+		 *
+		 * @TODO: Handle NO_BACKSLASH_ESCAPES SQL mode.
+		 */
+		$value = stripcslashes( $value );
+
+		/*
+		 * 5. Translate datetime literals.
+		 *
+		 * Process only strings that could possibly represent a datetime
+		 * literal ("YYYY-MM-DDTHH:MM:SS", "YYYY-MM-DDTHH:MM:SSZ", etc.).
+		 */
+		if ( strlen( $value ) >= 19 && is_numeric( $value[0] ) ) {
+			$value = $this->translate_datetime_literal( $value );
+		}
+
+		/*
+		 * 6. Handle null characters.
+		 *
+		 * SQLite doesn't fully support null characters (\u0000) in strings.
+		 * However, it can store them and read them, with some limitations.
+		 *
+		 * In PHP, null bytes are often produced by the serialize() function.
+		 * Removing them would damage the serialized data.
+		 *
+		 * There is no way to store null bytes using a string literal, so we
+		 * need to split the string and concatenate null bytes with its parts.
+		 * This will convert literals will null bytes to expressions.
+		 *
+		 * Alternatively, we could replace string literals with parameters and
+		 * pass them using prepared statements. However, that's not universally
+		 * applicable for all string literals (e.g., in default column values).
+		 *
+		 * See:
+		 *   https://www.sqlite.org/nulinstr.html
+		 */
+		$parts = array();
+		foreach ( explode( "\0", $value ) as $segment ) {
+			// Escape and quote each segment.
+			$parts[] = "'" . str_replace( "'", "''", $segment ) . "'";
+		}
+		if ( count( $parts ) > 1 ) {
+			return '(' . implode( ' || CHAR(0) || ', $parts ) . ')';
+		}
+		return $parts[0];
+	}
+
+	private function translate_pure_identifier( WP_Parser_Node $node ): string {
+		$token = $node->get_child_token();
+
+		if ( WP_MySQL_Lexer::DOUBLE_QUOTED_TEXT === $token->id ) {
+			$value = substr( $token->value, 1, -1 );
+			$value = str_replace( '""', '"', $value );
+		} elseif ( WP_MySQL_Lexer::BACK_TICK_QUOTED_ID === $token->id ) {
+			$value = substr( $token->value, 1, -1 );
+			$value = str_replace( '``', '`', $value );
+		} else {
+			$value = $token->value;
+		}
+
+		return '`' . str_replace( '`', '``', $value ) . '`';
```

Comment:
Is doubling the backticks the only rule we have to follow? Is there a way to derail this? What would happen if there's a backslash backtick combination? Or something like a LTR UTF-8 character inside the string?

Reply:
I believe so, yes. In SQLite literals, there are no Unicode escape sequences or anything like that. You can only escape the quote, and only by doubling it. A few quick tests:

```sql
SELECT 'a\b'; -- a\b
SELECT 'a'b'; -- syntax error
SELECT 'a\'b'; -- syntax error
SELECT 'a\''b'; -- a\'b
```

And here is a quick test with LTR:

```sql
SELECT 'a‎b' == 'ab'; -- 0 (because the first literal has LTR between 'a' and 'b')
```

Reply:
That sounds promising, thanks! Just to be extra diligent here – is there anything in the SQLite test suite related to escaping that catches your attention?

Reply:
@adamziel As mentioned in https://github.com/Automattic/sqlite-database-integration/pull/1#discussion_r1946737335, I'll create a ticket for this, will add a larger test case, and we can discuss further there.